src/main/java/org/apache/mesos/scheduler/repair/constrain/TimedLaunchConstrainer.java

```diff
@@ -0,0 +1,38 @@
+package org.apache.mesos.scheduler.repair.constrain;
+
+import org.apache.mesos.Protos.Offer.Operation;
+import org.apache.mesos.offer.OfferRequirement;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Implements a {@link LaunchConstrainer} that requires a minimum number of seconds to elapse between launches, for
+ * rate-limiting purposes.
+ */
+public class TimedLaunchConstrainer implements LaunchConstrainer {
+    private AtomicLong lastLaunch;
+    private long minDelaySeconds;
+
+    /**
+     * Create a new constrainer with the given required minimum delay between recovery operations.
+     *
+     * @param minDelaySeconds Minimum number of seconds between each launch
+     */
+    public TimedLaunchConstrainer(long minDelaySeconds) {
+        this.minDelaySeconds = minDelaySeconds;
+        this.lastLaunch = new AtomicLong(0);
+    }
+
+    @Override
+    public void launchHappened(Operation launchOperation) {
+        if (!canLaunch(null)) {
```

Comment:
So, I think you're thinking through the wrong case. The issue is if `canLaunch` returns **true**, and then `lastLaunch` is updated. In this case, we can observe a launch that shouldn't have happened. The issue is that we don't have any atomicity guarantee between these 2 methods. I try to avoid creating implicit threading demands on APIs, so that future concurrent code won't randomly break.

The issue here is that multiple threads can be making the `canLaunch` -> `launchHappened` paired calls. There are interleavings that would invalidate this logic, and I believe that when we can detect a race condition, we should throw something.

There are 3 solutions:

1. The solution I implemented, which detects the bug without preventing future concurrency.
1. We could put a lock around the repairscheduler, which would prevent the bug but also prevent future concurrency.
1. I can remove this check, and the behavior will remain the same, although then the bug will simply become a silent issue.

I think the solution I chose is best because it detects the issue w/o reducing the future design space.